### ISSUE TITLE
Very slow git submodule checking in a Jenkins environment on Windows

### ISSUE THREAD

**Issue by mgedmin** [mgedmin/check-manifest#63]
When I added [git submodule checking](https://github.com/mgedmin/check-manifest/commit/effec6c72e91a3ee4812a4f6e9e554ebb6a89e28), my Jenkins job run time went from 10 minutes up to 1 hour and 45 minutes.  *Ouch*.

It turns out that every invocation of `git submodule ...` [takes 50 seconds](https://jenkins.gedmin.as/job/check-manifest-on-windows/430/console) -- but only inside Jenkins.  I cannot reproduce this from a command shell on the same machine.

I suspect a different %PATH% setting might be the cause.  Perhaps Windows systems take a long time not finding executables on a long %PATH%?

(Meanwhile, I implemented a [workaround](https://github.com/mgedmin/check-manifest/commit/507ee2e90144e8e47ac64eea84119e3498b8eee6), but that still leaves me a job that takes 30 minutes instead of the old 10 minutes.)

**Comment by mgedmin** [comment]
`TestGit.test_get_versioned_files_with_git_submodules` takes 4 minutes 15 seconds to run.  Times the number of different Python interpreters (5) and we get those extra 20 minutes.

**Comment by mgedmin** [comment]
It's not %PATH%.  :-(

**Comment by mgedmin** [comment]
New hypothesis from https://github.com/mgedmin/check-manifest/pull/73#issuecomment-243113796: my Jenkins has a pretty ancient Git for Windows from ~~1993~~ 2013: 1.8.4-preview20130916.  Let's install something more modern and see if that speeds things up.

**Comment by lorengordon** [comment]
very nice, i shall follow along here now, because snarky comments about windows cheers me up on monday morning

**Comment by mgedmin** [comment]
I've removed the old git, installed a new git, and rebooted the VM to apply pending security updates (how does one set up unattended reboots in Windows 2012 Server anyway?).

My Jenkins hasn't realized yet that the job it was in the middle of running was killed by the reboot.
`¯\_(ツ)_/¯`

**Comment by mgedmin** [comment]
Oops I lost my nice /etc/motd where I'd noted where my Jenkins logs live :/

**Comment by mgedmin** [comment]
The worker cannot connect to the master because

*drumroll*
```
INFO: Server didn't accept the handshake: langas.gedmin.as is already connected to this master. Rejecting this connection.
```
:100:

**Comment by lorengordon** [comment]
Open the control panel, and navigate to this page, long as no one is logged on when it goes to auto-install updates, it should also automatically reboot...

![image](https://cloud.githubusercontent.com/assets/8457307/18052831/84b81b46-6dca-11e6-9c81-fea91c747727.png)

**Comment by mgedmin** [comment]
The update is already set to "install automatically (recommended)", with the maintenance window at 3 AM.  Nevertheless every time I connect (which happens once every few months), if I stay connected for more than 5 minutes, I'm likely to get the blue "your computer needs to be restarted to apply updates" prompt.

**Comment by mgedmin** [comment]
Meanwhile I restarted the Jenkins master and it's building now: https://jenkins.gedmin.as/job/check-manifest-on-windows/690/console

**Comment by lorengordon** [comment]
windows :gun: 

Windows update is pretty broken, from a configuration perspective. I've used this resource before, to try to figure out how to configure it:
- [All Windows Update registry keys](https://technet.microsoft.com/en-us/library/Dd939844(v=WS.10).aspx)

**Comment by mgedmin** [comment]
So far instead of being faster it seems to be slower?  7 minutes gone and it hasn't finished test_get_versioned_files_with_git_submodules (tests.TestGit).

**Comment by mgedmin** [comment]
Ooh,

> NoAutoRebootWithLoggedOnUsers
> 1 = Logged-on user can decide whether to restart the client computer.
> 0 = Automatic Updates notifies the user that the computer will restart in 15 minutes.

so when I'm done fiddling I just click [x] on my rdesktop client, which probably leaves me logged in.  Perhaps that prevents unattended reboots?

**Comment by mgedmin** [comment]
Another annoyance: when I launch Git Bash it starts in /.

The desktop icon properties say "Start in: %HOMEDRIVE%%HOMEPATH%" and Target is "C:\Git\git-bash.exe --cd-to-home" so I've no idea why it starts in /.

If I run `/c/git/git-bash.exe --cd-to-home` from another bash shell, it starts in ~ (which is c:/Users/mg).

**Comment by lorengordon** [comment]
Ahh, yeah, that creates a disconnected session, which will prevent the reboot. There are also ways of terminating those sessions via other settings...

**Comment by lorengordon** [comment]
I use powershell and posh-git, rather than git-bash, so no ideas on that one...

**Comment by mgedmin** [comment]
Why is my jenkins-slave.exe running as SYSTEM?  Is that normal?  Is that good?

**Comment by lorengordon** [comment]
windows... so probably normal but not good

**Comment by mgedmin** [comment]
Meanwhile the test's been stuck for 19 minutes.  I can see a sh.exe eating 40% CPU, but I can't figure out how to see the process three, or the full command line with Task Manager.

**Comment by lorengordon** [comment]
setting the user, note step 5, from https://wiki.jenkins-ci.org/display/JENKINS/Step+by+step+guide+to+set+up+master+and+slave+machines:


> If you want the service to run on start-up of the slave machine do the following (Windows only directions):
> 1. In the Slave agent program running on your slave machine,  click File --> Install as Windows Service. Note that this feature requires ".Net Framework 3.5"
> 2. Start, type Services and Select the Services program.
> 3. Find Jenkins Slave in the list, Double click to open.
> 4. Select Startup type --> Automatic.
> 5. Go to the Log On tab, change the Log on as to a user of your choice (Special user account Jenkins recommended).
> 6. Make sure that auto login is set for the slave machine for the user account, then the VM (or physical computer) should connect and be available when needed.

**Comment by mgedmin** [comment]
Process Explorer to the rescue!  It's
```
c:\git\usr\bin\sh.exe c:\git\mingw64/libexec/git-core\git-submodule add ../repo3 sub3
```
and it's been running for 24 minutes already.

Hmm.

**Comment by mgedmin** [comment]
Giving up for today.

**Comment by lorengordon** [comment]
hrm... i've sometimes noticed odd issues on windows if i do not specify .git as the repo extension?

**Comment by lorengordon** [comment]
i.e perhaps `git submodule add ../repo3.git sub3`?

**Comment by mgedmin** [comment]
(a) adding .git may be better style, but (b) the test repositories created in the test don't currently name the directories with .git, and (c) it works fine on Appveyor and worked fine on my Jenkins before I upgraded git...

TODO:
- [ ] reconfigure Jenkins service to use a dedicated user account
- [ ] figure out how one adds a user account on Windows
- [ ] play with git submodules manually on the command prompt, see what happens
- [ ] fix the startup directory problem
- [ ] see if Git for Windows has an issue tracker somewhere; report the startup directory problem
- [ ] rethink poor life choices

**Comment by mgedmin** [comment]
Oh right those are non-bare repositories.

Maybe I want to run `git submodule add ../repo3/.git` and see what happens.

**Comment by mgedmin** [comment]
I added `c:\git\usr\bin` to `%PATH%` in my Jenkins node config and now the tests succeed in 2 minutes 25 seconds!  :fireworks:

AFAICT the old git for windows put everything in `c:\git\bin`, which was already in my `%PATH%`, but the new one put most of the Unix utils in `c:\git\usr\bin`, which wasn't.  I noticed this because an unrelated build started failing due to not finding `true.exe` on `%PATH%`, and then when I fixed that and started debugging the check-manifest getting stuck, hey, it's not getting stuck any more!
